controller/resource.go

```diff
@@ -196,79 +390,112 @@ type ServiceInfo struct {
 		return s.Delete()
 	}
 
 	return nil
 }
 
-// Upsert creates a token for the task if one doesn't already exist
+// Upsert creates a service policy and token for the task if one doesn't already exist
 // and updates the secret with the contents of the token.
 func (s *ServiceInfo) Upsert() error {
+	opts := &api.QueryOptions{Partition: s.QName.Partition(), Namespace: s.QName.Namespace()}
+
+	// policy
+
+	currPolicy, _, err := s.ConsulClient.ACL().PolicyReadByName(s.policyName(), opts)
+	if err != nil && !IsACLNotFoundError(err) {
+		return fmt.Errorf("reading service policy: %w", err)
+	}
+
+	// if the policy doesn't already exist, create it.
+	if currPolicy == nil {
+		if err = s.createServicePolicy(); err != nil {
+			return fmt.Errorf("creating service policy: %w", err)
+		}
+	}
+
+	// token
+
 	currSecret, err := s.upsertSecret()
 	if err != nil {
 		return fmt.Errorf("upserting secret: %w", err)
 	}
 
-	currToken, _, err := s.ConsulClient.ACL().TokenRead(currSecret.AccessorID, nil)
+	currToken, _, err := s.ConsulClient.ACL().TokenRead(currSecret.AccessorID, opts)
 
-	if err != nil && !isACLNotFoundError(err) {
+	if err != nil && !IsACLNotFoundError(err) {
 		return fmt.Errorf("reading existing token: %w", err)
 	}
 
 	// If there is already a token for this service in Consul, exit early.
 	if currToken != nil {
-		s.Log.Info("token already exists; skipping token creation", "id", s.ServiceName)
+		s.Log.Info("token already exists; skipping token creation", "id", s.QName)
 		return nil
 	}
 
 	// Otherwise, create one.
-	err = s.createServiceToken(currSecret)
-	if err != nil {
-		return fmt.Errorf("updating service token: %w", err)
+	if err = s.createServiceToken(currSecret); err != nil {
+		return fmt.Errorf("creating service token: %w", err)
 	}
 
 	return nil
 }
 
-// Delete removes the token for the given ServiceInfo.
+// Delete removes the service policy and token for the given ServiceInfo.
 func (s *ServiceInfo) Delete() error {
+	opts := &api.WriteOptions{Partition: s.QName.Partition(), Namespace: s.QName.Namespace()}
+
 	for _, token := range s.ServiceState.ACLTokens {
-		_, err := s.ConsulClient.ACL().TokenDelete(token.AccessorID, nil)
+		_, err := s.ConsulClient.ACL().TokenDelete(token.AccessorID, opts)
 		if err != nil {
 			return fmt.Errorf("deleting token: %w", err)
 		}
-		s.Log.Info("token deleted successfully", "service", s.ServiceName)
+		s.Log.Info("token deleted successfully", "service", s.QName)
+	}
+
+	for _, policy := range s.ServiceState.ACLPolicies {
+		_, err := s.ConsulClient.ACL().PolicyDelete(policy.ID, opts)
+		if err != nil {
+			return fmt.Errorf("deleting policy: %w", err)
+		}
+		s.Log.Info("policy deleted successfully", "policy", s.policyName())
 	}
 
 	return nil
 }
 
+// GetNamespace returns the namespace that the service belongs to.
+// It returns the empty string if namespaces are not enabled.
+func (s *ServiceInfo) GetNamespace() string {
```

Comment:
nit: `GetNamespace()` -> `Namespace()` to be consistent (and clarify this isn't fetching the namespace). Or, we could access `s.QName.Namespace()` directly, which I see in some places as well.

Reply:
Agreed. I have renamed `GetNamespace()` to `Namespace()`. This is needed to allow `ServiceInfo` to implement the `Namespace()` func in the `Resource` interface. Which is now used by the controller to `ReconcileNamespaces()` outside of the `List()` function.